src/lock_free/mwcas_descriptor.cpp

```diff
@@ -191,37 +205,40 @@ MwCASDescriptor::MwCASInternal(  //
   }
 
   const auto succeeded = (cur_stat == kSucceeded);
+  auto all_cnts_zero = true;
   if (succeeded) {
     for (size_t i = 0; i < target_cnt_; ++i) {
       auto &target = targets_[i];
       const auto ver = (target.old_val.load(kRelaxed) + kVersionUnit) & kVersionMask;
-      Finalize(base_addr, target, (target.new_val | ver));
+      all_cnts_zero &= Finalize(base_addr, target, (target.new_val | ver));
     }
   } else {
     for (size_t i = 0; i < target_cnt_; ++i) {
       auto &target = targets_[i];
       const auto val = (target.old_val.load(kRelaxed) + kVersionUnit) & kVerAndValMask;
-      Finalize(base_addr, target, val);
+      all_cnts_zero &= Finalize(base_addr, target, val);
     }
   }
 
+  if (all_cnts_zero) {
+    _tls.reset(this);
+  }
+
   return succeeded;
```

Comment:
- `&=` はビット演算子なので、bool値に対して使うのは実はよくないです。
- 記述子を解放する（TLSで保持する or GCに加える）のは所有者の役目なので、こちらでは何もしなくてよいです。代わりに、他のスレッドから参照されているかのフラグだけ返します。

```suggestion
  uint64_t ref_cnt{};
  if (succeeded) {
    for (size_t i = 0; i < target_cnt_; ++i) {
      auto &target = targets_[i];
      const auto ver = (target.old_val.load(kRelaxed) + kVersionUnit) & kVersionMask;
      ref_cnt += Finalize(base_addr, target, (target.new_val | ver));
    }
  } else {
    for (size_t i = 0; i < target_cnt_; ++i) {
      auto &target = targets_[i];
      const auto val = (target.old_val.load(kRelaxed) + kVersionUnit) & kVerAndValMask;
      ref_cnt += Finalize(base_addr, target, val);
    }
  }

  return {succeeded, ref_cnt > 0};
```